### ISSUE TITLE
Create a minimal Fluvio interface definition in preparation for open-source release

### ISSUE THREAD

**Issue by nicholastmosher** [infinyon/fluvio#277]
Brainstorm and document a minimal Fluvio user interface in terms of a generic domain model. Elements of the interface should not be tied to a specific technology, but rather should express the actions a user may want to take when working with Fluvio, regardless of the client they're using. The goal of this is to create a unified mental model for interacting with Fluvio which is shared among all of the client interfaces. This is the first step in each of #245, #240, and #246.

## Goals

By the end of this issue and discussion, we should have a relatively full picture of the functionality which an end user will have access to. We should expect that any existing Fluvio interface which is not described in this issue will be made private and unavailable to the user before our open-source release. We can use existing examples as inspiration for the scope of functionality that should ultimately exist in this issue's interface definition, though we may clarify, minimize, or rearrange that interface if it seems helpful.

## Examining existing functionality

I went through one of our end-to-end examples (CDC) and picked out all of the usages of the older `flv-client` crate. I'm going to list the types and methods below, and from the code I'm going to write up a plain-English description of what is going on in terms of an end-user perspective. Please let me know in the comments if I am making any assumptions or if my interpretations of any of these are incorrect.

### Producer Example

- `ScClient`
  - `.topic_metadata`: I would like to be able to discover existing topics in my cluster
  - `.create_topic`: I would like to be able to create a new topic to send events to
  - `.find_replica_for_topic_partition`: If I know my topic and partition, I would like to find out which SPU is responsible for streaming that data (SPU replica leader)
- `ScConfig`
  - `::new`: I would like to set up any settings needed to connect to a cluster
  - `.connect`: I would like to connect to a cluster using these settings
- `SpuReplicaLeader`: Acts as a reference to a particular SPU, the one that is in charge of my topic and partition
  - `.fetch_logs_once`: I would like to fetch a range of logs all at once, not as a stream
  - `.send_record`: I would like to append a message to the topic and partition managed by this SPU leader
- `ReplicaConfig`
  - `::Computed`: I would like to specify how my topic will be replicated using a partition count and replica count
- `FetchLogOption`: I would like to configure some parameters for how my fetch will behave
- `FetchOffset`: I would like to specify a log index to start or end a fetch

### Consumer Example

- `SpuReplicaLeader`
  - `.fetch_logs`: I would like to watch for new events arriving on my topic and partition

## Domain Terminology

Here I am listing some terms from the streaming domain that I have seen around the Fluvio codebase and project. I will also list some other terminology which either a) we use but is not commonplace, or b) is commonplace but we don't use it. We may want to brainstorm if there are any ways we can better align our own terminology with the wider streaming domain terminology.

Existing terminology in Fluvio
- a "Fetch" is a one-time batch delivery of events from an offset and a range
- a "Stream" is a continuous delivery of events in real-time
- a "Topic" is a user-defined category of events
- a "Partition" is a fragment of a Topic. Different partitions may be processed by different SPUs.
- a "Replica" refers to a full copy of a Topic, including all of the topic's Partitions. If a Topic has a replication factor of 3, then there will be three copies of all of the data in that topic, and therefore three copies of that topic's partitions.
- an "Offset" is an index from the front or back of a partition which locates an event
- (proposed) a "Range" specifies a batch of events. Types of ranges may include:
  - FromBeginning (until offset)
  - (from offset) UntilEnd
  - BetweenOffsets

More general streaming terminology (not seen in Fluvio yet)
- a "Partition Key" is a field that uniquely identifies an event and is used to determine which Partition the event will be processed by

## Desired functionality

Below, we'll list functionality point-by-point that we want included in the release. Each piece of functionality should trace back to a user's intent.

- ### Set up Fluvio
  - User Intent: As a new user, I want to be able to get started with Fluvio on my local machine
  - Applicable clients: CLI

- ## Typical user functionality

  - [x] Initialize Client
    - User Intent: I want to connect to my Fluvio cluster with default or custom settings
    - Applicable clients: CLI, Rust, Node
    - API guidelines:
      - There should be a client object type which is given to the user upon successful connection and authentication to a Fluvio cluster. If the connection fails, the user should not have access to a Fluvio client object. This way, subsequent API functionality such as producing and consuming is gated upon successful connection and users cannot accidentally try to produce or consume without passing this step.
      - Example implementation: Rust client initialization looks like `async Fluvio::connect() -> Result<Fluvio, FluvioError>`

  - [x] Produce messages
    - User Intent: I want to be able to send messages from my application to a topic
    - Applicable clients: CLI, Rust, Node
    - API guidelines:
      - There should be a way to construct a producer object from the client object. A producer is bound to a topic, but may send messages to various partitions.
      - Example: Rust producer is created like `async Fluvio::topic_producer(topic: String) -> Result<TopicProducer, FluvioError>`. Note that this type of producer is named to reflect that it is bound to a topic.
      - There should also be a way to construct a producer object directly, without a Fluvio client object. In this case, the constructor should assume all-default settings.
      - Example: The Rust client package has a global producer constructor that looks like `async fluvio::producer(topic: String) -> Result<TopicProducer, FluvioError>`

  - [x] Consume messages
    - User Intent: I want to be able to consume messages from one of my topics
    - Applicable clients: CLI, Rust, Node
    - API guidelines:
      - There should be a way to construct a consumer from the client object. The consumer should be bound to a topic and a specific partition.
      - Example: Rust producer is created with `async Fluvio::partition_consumer(topic: String, partition: i32) -> Result<PartitionConsumer, FluvioError>`. Note that this type of consumer is named to reflect that it is bound to a partition.
      - There should also be a way to construct a consumer object directly, without a Fluvio client object. In this case, the constructor should assume all-default settings.
      - Example: In rust there is the following consumer producer: `async fluvio::consumer(topic: String, partition: i32) -> Result<PartitionConsumer, FluvioError>`

**Comment by nicholastmosher** [comment]
Brainstorming ideas for topic consumer interface (pseudocode):

```rust
fn main() {
    let fluvio = Fluvio::new(); // or some other constructor

    // Note: topic_consumer is an external "streaming iterator" that yields event items
    // and keeps track of the current offset internally
    let topic_consumer: TopicConsumer = fluvio.consume_topic("topic-name");

    while let Some(event: FluvioEvent) = topic_consumer.next().await {
        // Process event
        let structured_event: MyEvent = event.try_into()?;
    }
}

// A wrapper around the user's event type which has some sort of easy coercion
// into the user's type, but also provides access to Fluvio metadata such as offsets, etc.
struct FluvioEvent<T> {
    user_event: T,
    offset: ...
}
```

**Comment by sehz** [comment]
`consume_topic` seems redundant since you can only consume from topic

**Comment by nicholastmosher** [comment]
So a disclaimer to all of this is that I haven't ever used the existing Fluvio API. I'm starting to go back now and look at the examples to see how things were set up before. But so if you can only consume from a topic, what are the things that a producer needs to know? I had assumed that a producer would just send messages to a topic

**Comment by sehz** [comment]
also, take look at current API which is different from previous API. 

Also, have to think of how many API we want to expose.  Curren API workflow is start from  ClusterSocket, then
- connect to cluster
- then specialized API such as Admin, Consumer and Producer.

example:
```
let cluster_config =  ConfigFile::load()?;
let mut cluster = ClusterSocket::connect(cluster_config.clone()).await?;
let mut consumer = cluster.consumer(("topic",0)).await?;
....consume....
```

Alternative is to have top 3 Cluster objects developer can interact with:
- Consumer
- Produce
- Admin

example:
```
let cluster_config =  ConfigFile::load()?;
let mut consumer = Consumer::connect(cluster_config,("topic",0)).await?;
// or alternatively
let out consumer = Consumer::subscribe("topic",0)).await?;     // use default profile
... consume...
```

**Comment by ajhunyady** [comment]
I have some questions that may impact the APIs

Is the client connection to the cluster shared?
      let mut connection = ClusterSocket::connect(cluster_config.clone()).await?;

If each connection is independent, then it makes sense to go directly to:
       let mut consumer = Consumer::connect(cluster_config,("topic",0)).await?;

**Comment by sehz** [comment]
For first case, connection to a cluster is shared.  Consumer leverages metadata from cluster to search topics.

In a second case, connection is independent.

**Comment by ajhunyady** [comment]
Not sure I understand what metadata that be leveraged.

**Comment by sehz** [comment]
consumer uses metadata to search for topics

**Comment by nicholastmosher** [comment]
@sehz @ajhunyady I have updated the description with an "Examining Existing Functionality" section. Could you please read through it and tell me if I am on track or have any misconceptions?

**Comment by nicholastmosher** [comment]
Closed by #290 and #292
